File: labs/lab2.3/cpp/TextContainer.cpp
#include "headers/TextContainer.h"
#include "headers/TextLine.h"
#include "stdexcept"

TextContainer::TextContainer() : lines{} {};
TextContainer::TextContainer(const string &line) : lines{TextLine(line)} {}
TextContainer::TextContainer(const TextLine &line) : lines{line} {}
TextContainer::TextContainer(const vector<string> &lines) {
    for (const string &i : lines) this->lines.push_back(TextLine(i));
}
TextContainer::TextContainer(const vector<TextLine> &lines) : lines(lines) {}
TextContainer::TextContainer(const TextContainer &other) : lines(other.lines) {}
void TextContainer::Append(const string &line) {
    this->lines.push_back(TextLine(line));
}
void TextContainer::Append(const TextLine &line) {
    this->lines.push_back(line);
}
void TextContainer::Clear() {
    this->lines.clear();
}
void TextContainer::RemoveLine(int line) {
    if (line < 0 || line >= this->lines.size()) throw std::invalid_argument("Line is out of range");
    this->lines.erase(this->lines.begin() + line);
}
int TextContainer::LineCount() const { return this->lines.size(); }
int TextContainer::CharCount() const {
    int count = 0;
    for (TextLine i : this->lines) count += i.Length();
    return count;
}
float TextContainer::ConsonantRatio() const {
    int consonantCount = 0;
    for (TextLine i : this->lines) consonantCount += i.GetConsonantCount();
    return (float)consonantCount / this->CharCount();
}
TextLine TextContainer::GetShortestLine() const {
    TextLine shortest = this->lines[0];
    for (int i = 1; i < this->lines.size(); i++) {
        if (this->lines[i].Length() < shortest.Length()) shortest = this->lines[i];
    }
    return shortest;
}
void TextContainer::RemovePrefixWhiteSpace() {
    for (TextLine &i : this->lines) i.RemovePrefixWhiteSpace();
}
void TextContainer::RemovePostfixWhiteSpace() {
    for (TextLine &i : this->lines) i.RemovePostfixWhiteSpace();
}
void TextContainer::RemoveExtraSpaces() {
    for (TextLine &i : this->lines) i.RemoveExtraSpaces();
}
string TextContainer::ToString() const {
    if (this->lines.size() == 0) return "";
    string result = this->lines[0].ToString();
    for (int i = 1; i < this->lines.size(); i++) {
        result += '\n' + this->lines[i].ToString();
    }
    return result;
}
TextLine& TextContainer::operator[](int key) { return this->lines[key]; }

File: labs/lab2.3/csharp/TextContainer.cs
// Клас-контейнер, який є абстракцією тексту та складається з
// об’єктів класу-рядка та методів додавання рядка до тексту,
// видалення рядка з тексту, очищення тексту, отримання
// довжини найкоротшого рядка, повернення відсотка
// приголосних літер у тексті, заміни сукупності пробілів, що
// стоять поряд, одним пробілом, та видалення пробілів
// спочатку рядка та наприкінці.

namespace lab2_3;

class TextContainer : IWhiteSpaceRemover{
    private TextLine[] lines;

    public int LineCount {get => this.lines.Length;}
    public int CharCount {get {
        int count = 0;
        foreach (var i in this.lines) count += i.Length;
        return count;
    }}
    public TextLine this[int key] {get => this.lines[key];}

    public TextContainer() {
        this.lines = [];
    }
    public TextContainer(string line) {
        //creates a text container from one string object
        this.lines = [new TextLine(line)];
    }
    public TextContainer(TextLine line) {
        //creates a text container from one TextLine object
        this.lines = [line];
    }
    public TextContainer(string[] lines) {
        this.lines = new TextLine[lines.Length];
        for (int i = 0; i < lines.Length; i++) this.lines[i] = new TextLine(lines[i]);
    }
    public TextContainer(TextLine[] lines) {
        this.lines = lines;
    }
    public TextContainer(TextContainer other) {
        this.lines = other.lines;
    }

    public void Append(string line) => this.lines = [..this.lines,new TextLine(line)];
    public void Append(TextLine line) => this.lines = [..this.lines,line];
    public void Clear() {
        Array.Clear(this.lines);
        this.lines = [];
    }
    public void RemoveLine(int line) {
        if (line < 0 || line >= this.lines.Length) throw new IndexOutOfRangeException($"Line {line} is out of range");
        var asList = new List<TextLine>(this.lines);
        asList.RemoveAt(line);
        this.lines = asList.ToArray();
        asList.Clear();
    }
    public TextLine GetShortestLine() {
        TextLine shortestLine = this.lines[0];
        for (int i = 1; i < this.lines.Length; i++) {
            if (this.lines[i].Length < shortestLine.Length) shortestLine = this.lines[i];
        }        
        return shortestLine;
    }
    public IEnumerator<TextLine> GetEnumerator() {
        foreach (TextLine i in this.lines) yield return i;
    }
    public float ConsonantRatio() {
        int consonantCount = 0;
        foreach (var i in this.lines) consonantCount += i.Consonants;
        return (float)consonantCount / this.CharCount;
    }
    public void RemovePrefixWhiteSpace() {
        foreach (var i in this.lines) i.RemovePrefixWhiteSpace();
    }
    public void RemovePostfixWhiteSpace() {
        foreach (var i in this.lines) i.RemovePostfixWhiteSpace();
    }
    public void RemoveExtraSpaces() {
        foreach (var i in this.lines) i.RemoveExtraSpaces();
    }
    public override string ToString() => string.Join('\n',this.lines.Select(line => line.ToString()));

    public static TextContainer operator +(TextContainer left, TextLine right) => new TextContainer([..left,right]);
    public static TextContainer operator +(TextContainer left, string right) => new TextContainer([..left,new TextLine(right)]);
    public static TextContainer operator +(TextContainer left, TextLine[] right) => new TextContainer([..left,..right]);
    public static TextContainer operator +(TextContainer left, string[] right) {
        TextLine[] lines = new TextLine[left.lines.Length + right.Length];
        for (int i = 0; i < left.lines.Length; i++) {
            lines[i] = left.lines[i];
        }
        for (int i = 0; i < right.Length; i++) {
            lines[left.lines.Length + i] = new TextLine(right[i]);
        }
        return new TextContainer(lines);
    }
    public static TextContainer operator +(TextContainer left, TextContainer right) => new TextContainer([..left.lines,..right.lines]);
}